File: LinkedList/Singly/Prepend.js
//Add Node at the beginning of the Linked list

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  prepend(data) {
    const newNode = new Node(data);

    newNode.next = this.head;
    this.head = newNode;
    this.length++;

    if (this.tail === null) {
      this.tail = newNode;
    }
  }

  print() {
    let current = this.head;
    let result = "";
    while (current) {
      result += `${current.data}` + `${current.next !== null ? " -> " : ""}`;
      current = current.next;
    }
    return result;
  }
}

const list = new LinkedList();

list.prepend(10);
console.log(list.print());
list.prepend(20);
console.log(list.print());
list.prepend(44);
console.log(list.print());
list.prepend(35);
console.log(list.print());

File: LinkedList/Singly/Python/Insertion/prepend.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def prepend(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node
        self.length += 1

        if self.tail is None:
            self.tail = node

    def print(self):

        current = self.head
        result = ""
        while current:
            result += f"{current.data}" + ("->" if current.next else "")
            current = current.next
        
        print(result)


if __name__ == "__main__":

    linked_list = LinkedList()

    linked_list.prepend(10)
    linked_list.prepend(20)
    linked_list.prepend(30)
    linked_list.prepend(40)

    linked_list.print()


    
